### ISSUE TITLE
Showing <remarks> in QuickInfo results in very large tooltips for Xamarin

### ISSUE THREAD

**Issue by sandyarmstrong** [dotnet/roslyn#42606]
**Version Used**: Version 16.6.0 Preview 2.0 [29918.199.master]

**Steps to Reproduce**:

1. Create new Xamarin.iOS single view project.
2. Open `ViewController.cs`.
3. Hover over `UIViewController`

**Expected Behavior**:

<img width="570" alt="image" src="https://user-images.githubusercontent.com/155076/77126638-0b974600-6a07-11ea-8d66-22ab04c0011d.png">

**Actual Behavior**:

<img width="2353" alt="image" src="https://user-images.githubusercontent.com/155076/77126657-1f42ac80-6a07-11ea-9e50-d85e43bc39cb.png">

---

This is the result of the (otherwise very useful!) https://github.com/dotnet/roslyn/pull/42135.

Mono and Xamarin historically make extensive use of `<remarks>` in documentation, and that information is typically kept intact for offline documentation (via the MonoDoc tool, see `Help -> API Documentation` on VSmac). Apparently, .NET Framework BCL documentation strips out `<remarks>` on user machines, but the data is available in online documentation.

VSmac has not yet inserted this week's Roslyn in order to avoid this issue on very basic types like `System.Console` and boolean literals:

![image](https://user-images.githubusercontent.com/155076/77126773-83fe0700-6a07-11ea-8db7-780a2e224f9c.png)

(Of note, the above behavior only occurs for Mono-based projects on Mac, but that includes Xamarin.iOS, etc)

---

To be fair, it has always been possible to make unhelpfully large QuickInfo tooltips even before this change. While I feel that we do need to seriously address the problem over large tooltips, that should probably be a separate issue.

For now, we should decide how to move forward on showing `<remarks>` content. Some proposals:

1. **Remove the feature entirely.**
   This would be a bummer, as it seems generally useful. But maybe it's the right thing to do if QuickInfo UX needs some new design work.
2. **Add an option for this feature, and disable it by default.**
   This might be the best default behavior, but many people would never discover the option.
3. **Add an option for this feature, enable it by default, but allow it to be disabled on a per-buffer level**
   This would require additional work from the Xamarin and VSmac teams to keep it turned off for Mono-based workspaces/buffers/whatever. These users would miss out on the feature and have no way to force it on (unless the IDE added an additional option for that). I'm not personally familiar with how to implement this for Xamarin on VSwin.
4. **Disable the feature for some specific namespaces, such as `System.*`, and perhaps `UIKit.*` and other Xamarin namespaces**
   Seems reasonable for most Visual Studio users, but doesn't help scenarios where namespaces we don't know about have similarly lengthy `<remarks>`.
5. **Truncate content to some "reasonable maximum"**
   This seems problematic for the case of boolean literals in Mono. It is not especially helpful to show anything beyond a simple tooltip when mousing over `false`. And conversely, I would not want to over-truncate reasonable documentation.

Personally, I'd prefer 2 or maybe 3, to be followed by more general work on what to do about too-long QuickInfo content (regardless of `<remarks>`).

Prior art on huge tooltips: https://github.com/dotnet/roslyn/issues/21493

**Comment by CyrusNajmabadi** [comment]
Tagging @jinujoseph for visibility.  We will have to do something here, though exact short term and medium term fixes are TBD.

**Comment by jasonmalinowski** [comment]
> Add an option for this feature, and disable it by default.

That might also be a prudent option right away while we figure out a better setup. Then change the default once we did figure out something better.

**Comment by jinujoseph** [comment]
interesting, Sam did bring this concern in design meeting. Short term I think we should revert or put the feature behind experiment flag.

**Comment by sharwell** [comment]
I'm leaning back towards option 4 (excluding **System.* **). It's possible we just hit an outlier that would not be frequent in practice.

**Comment by jasonmalinowski** [comment]
I had one more idea for a very short term mitigatoin: what about limiting to only symbols from source? The challenge with symbols from metadata is they can't edit their <remarks> down, they're just stuck with it. At least things from source they have a workaround?

**Comment by sandyarmstrong** [comment]
@jasonmalinowski yeah that's an intriguing approach, too.

Here's one more idea: pick some arbitrary acceptable `<remarks>` length, and only show them if they are under that length. Could seem unpredictable to users in 16.6 but it would just be a stopgap until we solve the more general problem of long QuickInfo content (and we could put priority on that for 16.7).

**Comment by sandyarmstrong** [comment]
Also pinging @davidortinau (Xamarin PM) here so he can weigh in after .NET Conf: Xamarin.

**Comment by davidortinau** [comment]
Hi all, truncating sounds like a confusing experience for developers. I would much prefer to see this change reverted since it introduces a new and breaking behavior that wasn't foreseen. Can we please do that and come up with a solution that works for everyone?

**Comment by CyrusNajmabadi** [comment]
> Hi all, truncating sounds like a confusing experience for developers.

Why would this be the case?

UI's truncate large pieces of information all the time.  Seems super common and really easy for people to grok...

**Comment by sharwell** [comment]
We discussed this in the design review meeting and want to take the following approach:

1. Exclude **System.\***, which is what we intended to do from the start
2. Limit the number of _paragraphs_ that can be shown from remarks, which allows for reasonably long content provided it is organized in a way that isn't overwhelming to present

**Comment by sandyarmstrong** [comment]
> Why would this be the case?

1. It's an overwhelming amount of data that distracts a user from their work. The beginning of a treatise on booleans or UIViewControllers is probably not that helpful.
2. There is no way to "expand" the truncated data, so if the truncated data _does_ seem helpful, it's not clear where to go to get the rest of it.
3. What do our UX designers have to say about increasingly large tooltips in the IDE?

> Limit the number of paragraphs that can be shown from remarks, which allows for reasonably long content provided it is organized in a way that isn't overwhelming to present

This is also problematic. Then it's not even clear that the content is truncated, so it may terminate in what feels like a really random place.

---

I think we just need to realize that many <remarks> were not written with this in mind. Xamarin's certainly weren't. If we're not turning this off by default or disabling it completely, we need to add more exclusions for Xamarin until the more general issue of long QuickInfo content is addressed.

**Comment by sandyarmstrong** [comment]
And from Xamarin.Android, here's a before:

![image](https://user-images.githubusercontent.com/155076/77582562-d3778380-6e9c-11ea-99cb-948468417c4f.png)

And after:

![image](https://user-images.githubusercontent.com/155076/77582605-decaaf00-6e9c-11ea-8c92-0276150347b0.png)

Aside from the issue of increased width, this tooltip is now dominated by licensing information. This was clearly intended for a page of documentation and not inline help.

**Comment by sandyarmstrong** [comment]
An exhaustive list of namespaces provided by Xamarin can be found here:

* https://github.com/xamarin/apple-api-docs/tree/master/en (very long list)
* https://github.com/xamarin/android-api-docs/tree/master/docs
* https://github.com/xamarin/Xamarin.Forms-api-docs/tree/master/docs (this one's easy, it's `Xamarin.Forms.*`)

Here's a Xamarin.Forms example where it's clear how inappropriate the content is out of the context of a doc viewer or doc website.

Before:

<img width="379" alt="image" src="https://user-images.githubusercontent.com/155076/77583494-6a910b00-6e9e-11ea-928a-a5bedc471132.png">

After:

<img width="1081" alt="image" src="https://user-images.githubusercontent.com/155076/77583522-7381dc80-6e9e-11ea-93f8-f179670071d0.png">

The last sentence implies content to follow which is just missing from the point of view of the developer.

**Comment by sharwell** [comment]
> This is also problematic. Then it's not even clear that the content is truncated, so it may terminate in what feels like a really random place.

It would end with a complete paragraph, followed by a separate line something like this:

> _Additional remarks were truncated..._

**Comment by sandyarmstrong** [comment]
@sharwell ah, thanks for the clarification, that is certainly better. But the Xamarin.Android and Xamarin.Forms examples above are presumably short enough to not be truncated and are still clearly problematic. We'll just have to add more exclusions.

**Comment by CyrusNajmabadi** [comment]
> There is no way to "expand" the truncated data, so if the truncated data does seem helpful, it's not clear where to go to get the rest of it.

Why not?  You can always 'goto def'.  We could alo have a hyperlink to just do this in the tooltip :)

**Comment by CyrusNajmabadi** [comment]
>  But the Xamarin.Android and Xamarin.Forms examples above are presumably short enough to not be truncated and are still clearly problematic. We'll just have to add more exclusions.

Sorry, can you clarify why they're 'clearly problematic'.  As a reader, those docs look great.  I'd much rather have the expanded version, with the useful information than the condensed form which tells me practically nothing...

**Comment by CyrusNajmabadi** [comment]
> Aside from the issue of increased width, this tooltip is now dominated by licensing information. This was clearly intended for a page of documentation and not inline help.

That's a problem with the page of documentation as well.  i would not want that at the top when reading the docs.  I would expect that to be at the bottom, with the docs organized from most relevant to least relevant info.

**Comment by sandyarmstrong** [comment]
> Sorry, can you clarify why they're 'clearly problematic'.

* "The last sentence implies content to follow which is just missing from the point of view of the developer."
* "this tooltip is now dominated by licensing information. This was clearly intended for a page of documentation and not inline help."

I'm not sure what else to say here. Before the tooltip showed what the doc developers wanted it to show. Now it does not. This is not the experience that was intended.

**Comment by sharwell** [comment]
I'm with Cyrus on this one. For both the Android.App example (https://github.com/dotnet/roslyn/issues/42606#issuecomment-604071976) and the Xamarin.Forms example (https://github.com/dotnet/roslyn/issues/42606#issuecomment-604076911), the new behavior seems like exactly what we want. The action item here is to update the documentation comments to not include website-specific content which will then ship as API documentation alongside the packaged assembly.

**Comment by CyrusNajmabadi** [comment]
> I'm not sure what else to say here. Before the tooltip showed what the doc developers wanted it to show. Now it does not.

But it shows me what i want to see as a developer.  This is why i want 'remarks'. They're useful docs.  They contain the information i need.  The summary is rarely good enough.   It's often little more than a repeat of what the name already tells me (or is so vague as to not be helpful).

These examples demonstrate it.  The existing docs are useless to me as the actual developer.  And i think quick info should cater to the developer, not the writer.  i.e. it's for the consumer's benefit, not hte producers.

**Comment by sandyarmstrong** [comment]
> the new behavior seems like exactly what we want

Who is "we" in this case? It is not what the Xamarin team wants and is not what the documentation writers intended.

> The action item here is to update the documentation comments to not include website-specific content which will then ship as API documentation alongside the packaged assembly.

There is a cost here to do this and still allow the API documentation to be available offline. I have no idea if this is realistic to occur in the 16.6 timeframe.

If this feature depends on documentation changes, it should not ship until those documentation changes are ready.

**Comment by CyrusNajmabadi** [comment]
> Who is "we" in this case? It is not what the Xamarin team wants

1. Why do your desires override my desires as the person consuming the docs?
2. If you really want this level of control over the doc experience, it may be the case that it's better to build your own doc experience.  As a VS user i find it super frustrating that a doc writer would tell me that i shouldn't look at the very docs they're writing even though i find it useful.

> If this feature depends on documentation changes, it should not ship until those documentation changes are ready.

I disagree.  For example, if you change the docs to remove htat useful information from the remarks, then it's just hurting me as a dev again.  Once again i have to go somewhere *else* to find exactly the info i need... :-/

**Comment by sandyarmstrong** [comment]
> Why do your desires override my desires as the person consuming the docs?

Neither of our desires matter much here. The Xamarin PM makes the call about what the best experience is for Xamarin developers.

**Comment by sharwell** [comment]
FWIW I agree with @sandyarmstrong that the CC AL license notice isn't what we want on that particular screen.

In either case, the _extraneous_ content for these cases is minimal to the point that it wouldn't be overwhelming, and perhaps wouldn't even be noticed by readers.

**Comment by CyrusNajmabadi** [comment]
>  The Xamarin PM makes the call about what the best experience is for Xamarin developers.

That's fine.  I think if xamarin wants a distinctive documentation experience, it would likely be best to maybe have that responsibility move over to xamarin team then.  That way you can cater this appropriately as best as you think makes sense for your audience.

**Comment by sharwell** [comment]
One easy way to resolve this from the Xamarin side prior to a more comprehensive approach is to filter the XML files in the SDK packages to not include `<remarks>` elements. This approach would match the approach used for .NET reference assemblies.

**Comment by tamlin-mike** [comment]
An option I haven't seen mentioned yet is to borrow how this works from IntelliJ. On first quick-info invocation you get the short version. If you invoke it again, you toggle (meaning you can toggle small-large-small-...) to a different view displaying all info. That second view is (IIRC), while floating, a full-fledged text view with scrollbar and so on.

**Comment by CyrusNajmabadi** [comment]
@sandyarmstrong is any work here still needed?

**Comment by sandyarmstrong** [comment]
> @sandyarmstrong is any work here still needed?

We disable showing remarks by default in VSMac. Users can turn it on if they want it.

I don't know if there are still issues with this on Windows. @Redth @jonpryor @dalexsoto is this still a problem on Windows in legacy Xamarin? In MAUI workloads? I know there are still holes in API docs being shipped in some workloads (https://github.com/xamarin/xamarin-macios/issues/13614).

**Comment by CyrusNajmabadi** [comment]
Closing out since we haven't eheard anything else and it's off by default for vsmac.
